[3 - 4단계 방탈출 예약 대기] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4327

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- [x]  한 메서드에 오직 한 단계의 들여쓰기만 한다.
- [x]  `else` 예약어를 쓰지 않는다.
- [ ]  모든 원시값과 문자열을 포장한다.
- [x]  한 줄에 점을 하나만 찍는다.
- [x]  줄여 쓰지 않는다.(축약 금지)
- [ ]  모든 엔티티를 작게 유지한다.
- [ ]  3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [ ]  일급 컬렉션을 쓴다.
- [ ]  getter/setter/property를 쓰지 않는다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 제로😄 듀이입니다.
항상 주말에 리뷰 요청드리게 되는 것 같아 더 감사한 마음으로 리뷰 받고 있습니다!
3 - 4단계도 잘 부탁드립니다🫡

### 테스트용 계정
- 아래 계정으로 테스트해주시거나, `data.sql`에서 더미데이터를 확인 후 테스트해주세요!
```
- admin 
  - id: admin@email.com
  - pw: password

- member (예약내역+대기내역 존재)
  - id:  member2@email.com 또는 member3@email.com
  - pw: 1234
```

### `ReservationInfo`
- 처음 예약 대기를 설계할 때에는 `Waiting`이 `Reservation`을 참조하도록 설계했습니다. 특정 예약에 대한 예약 대기라고 생각했기 때문입니다. 
- 그런데 `Waiting`이 알고자 하는 `Reservation`의 정보는 날짜, 시간, 테마뿐이었고, 그래서 필요한 정보만 `ReservationInfo`라는 값 객체를 만들어 `Waiting`에서 `@Embedded`로 참조하도록 했습니다. 
- `ReservationInfo`에 `reservationId`까지 포함된 것은 특정 예약에 대한 예약 대기라는 것을 명시해주기 위해서입니다.

### `Reservation` 논리삭제
- 예약을 취소하면 데이터베이스에서 예약을 물리삭제하지 않고 `Status`만 `CANCLED`로 변경해주어 논리삭제합니다. 
- 현재는 예약내역 조회 시 예약된 상태, 즉 `Status`가 `RESERVED`인 예약만 조회해오고 있습니다. 
만약 취소 상태의 내역까지 조회해와야 한다면 사용자의 내 예약 목록 조회에서는 '취소' 상태가 확인되지만, 관리자의 전체 예약 관리 목록에서는 '취소' 상태인지 확인되지 않기 때문에 우선은 주어진 클라이언트 코드를 변경하지 않고자 예약 상태인 것만 조회해오도록 해두었습니다.

### 동시성 문제
- 예약을 구현할 때에도 동시성 문제는 고려하지 않았기 때문에, 이번에 대기를 구현하면서도 동시성 문제는 고려하지 않고 구현했습니다.
그래도 동시성 키워드가 크루들 사이에서 이야기가 간간히 들려와 대략적으로 어떠한 키워드(컬럼 unique 설정, 비관적 락, 낙관적 락 등)로 학습하면 좋겠다는 것만 알아두었습니다.

### 테스트 코드
- 학습 목표에 `@DataJpaTest`, `@WebMvcTest`와 같은 슬라이스 테스트가 있었지만, 이미 `@SpringBootTest`로 많은 테스트가 구현되어 있어 기존 방식을 유지하여 서비스 테스트 코드를 추가했습니다.
- 추가된 기능에 대해 통합 테스트는 아직 많이 빈약합니다. 우선 프로덕션 코드에 대해 리뷰를 요청하고, 통합 테스트 코드를 추가하는 것이 낫겠다고 생각되어 이렇게 먼저 리뷰 요청드리는 것이니 양해 부탁드립니다🥲
 
### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

ljhee92 committed:
docs: 요구사항 명세 정리

ljhee92 committed:
feat: 클라이언트 코드 추가

ljhee92 committed:
docs: API 명세 추가

ljhee92 committed:
refactor: 쿠키 유효기간 수정

- 원활한 테스트를 위해 수정

ljhee92 committed:
refactor: 중복코드 제거 및 메서드명 수정

ljhee92 committed:
feat: 클라이언트 코드 수정(예약 대기 생성 요청)

ljhee92 committed:
feat: 예약 대기 요청 기능 구현

ljhee92 committed:
feat: 내 예약 목록 조회 시 예약 대기 목록 포함

ljhee92 committed:
feat: 예약 대기 취소 기능 구현

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
feat: 중복 예약 불가능 기능 구현

ljhee92 committed:
docs: 요구사항 명세 정리

ljhee92 committed:
feat: 어드민 예약 대기 관리 페이지 응답

ljhee92 committed:
feat: 클라이언트 코드 추가

ljhee92 committed:
test: 테스트 코드 수정

ljhee92 committed:
fix: 예약 대기 요청 시 과거 예약일 경우 불가하도록 수정

ljhee92 committed:
chore: 예약 대기 더미데이터 추가

ljhee92 committed:
feat: 어드민 예약 대기 목록 조회 기능 구현

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
feat: 어드민 예약 대기 취소 기능 클라이언트 코드 수정

ljhee92 committed:
feat: 예약 대기 승인 기능 구현

ljhee92 committed:
test: 테스트 코드 수정

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
test: 도메인 테스트 보강

ljhee92 committed:
test: 서비스 테스트 보강

ljhee92 committed:
test: 페이지 응답 테스트 추가

ljhee92 commented on src/main/java/roomescape/infrastructure/repository/WaitingRepository.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public interface WaitingRepository extends JpaRepository<Waiting, Long> {
+
+    long countByReservationInfo(ReservationInfo reservationInfo);
+
+    List<Waiting> findAllByMember(Member member);
+
+    boolean existsByReservationInfoAndMember(ReservationInfo reservationInfo, Member member);
+
+    List<Waiting> findAllByReservationInfo(ReservationInfo reservationInfo);
+
+    boolean existsByReservationInfo(ReservationInfo reservationInfo);
+}
```
- `countBy...`는 `rank` 계산을 위한 로직이에요
- 정렬은 DTO에서 해주고 있습니다.
- 이번 구현에서 동시성 문제는 고려하지 않았어요

asebn1 replied:
알겠습니다
추가로 참고하면 좋을자료 공유드려요

- https://jojoldu.tistory.com/516

ljhee92 replied:
공유해주신 자료 잘 읽어보았습니다 제로!
`exist~`가 `count~`보다 성능이 뛰어난 점은 알고 있었어요!
현재는 `count~`를 통해 rank의 갯수를 db에서 가져와서 +1 해주는 로직 때문에 사용한 것이었는데요.
만약 예약 대기의 갯수가 많아진다면 `count~`로 모든 예약 대기를 카운팅할 때 성능이 좋지 않을 것 같아요.
지금 생각나는 해결책으로는 쿼리로 rank의 가장 큰 값을 구해오는 방법 정도가 있을 것 같은데요!
제로는 어떻게 생각하시나요!?

asebn1 replied:
MAX를 조회해서 가장 큰 순번을 구하는 접근도 괜찮다고 생각합니다.
다만, 동시성 대응도 잘 잡아야 한다고 생각해요.
추가로, 자주 호출되지 않고 데이터가 엄청 많은 것이 아니라면 생각보다 DB에서 발생하는 성능가능성은 낮다고 생각합니다.
혹시나 성능때문에 로직이 어려워지는 케이스도 지양해야한다고 생각해요(트레이드오프겠죠?). 이 부분도 참고해주세요 :)

ljhee92 commented on src/main/java/roomescape/application/WaitingService.java:
```diff
@@ -0,0 +1,109 @@
+        return waitingRepository.existsByReservationInfo(reservationInfo);
+    }
+
+    public Waiting findFirstRankWaitingByReservationInfo(ReservationInfo reservationInfo) {
+        return waitingRepository.findAllByReservationInfo(reservationInfo)
+                .getFirst();
+    }
```
해당 메서드를 호출하기 전에 `ReservationService#cancelReservationById`에서 `Waitings`의 존재 여부를 확인하고 있기 때문에 처리해주지 않았어요
```java
@Transactional
    public void cancelReservationById(Long id) {
        Reservation reservation = findReservationById(id);
        reservation.cancel();
        ReservationInfo reservationInfo = ReservationInfo.create(reservation);
        if (waitingService.existsWaitings(reservationInfo)) { // 여기서 확인 중
            processWaitingToReservation(reservationInfo);
        }
    }

    private void processWaitingToReservation(ReservationInfo reservationInfo) {
        Waiting firstRankWaiting = waitingService.findFirstRankWaitingByReservationInfo(reservationInfo);
        ....
```

asebn1 replied:
findFirstRankWaitingByReservationInfo 내부에서도 최소한의 방어 로직을 추가해주시면 좋을 것 같아요.
호출 쪽에서 존재 여부를 미리 확인하더라도, 내부에서도 예외에 대비하는 것이 안전하고 코드라고 생각합니다.

ljhee92 replied:
동의합니다!
메서드를 호출해서 사용하는 쪽에서도 안전하게 사용할 수 있도록 방어적으로 로직을 추가하는 것이 더 좋을 것 같아요😄👍🏻
`exist~` + `findAll`로 처리하게 된다면 DB에 두 번 접근하게 되는 단점이 있고,
`findAll`만으로 처리하게 된다면 모든 대기를 찾아야 하는 단점이 있다고 생각해서 `Optional` 반환 + `findTop1~` 메서드로 성능도 개선해보았어요!
7f40943b 에서 반영해주었습니다.

asebn1 replied:
좋습니다 :)

ljhee92 committed:
fix: 트랜잭션 안정성 고려

ljhee92 commented on src/main/java/roomescape/application/ReservationService.java:
```diff
@@ -28,78 +31,84 @@ public class ReservationService {
+        ReservationInfo newReservationInfo = ReservationInfo.create(newReservation);
+
+        waitingService.updateWaitings(reservationInfo, newReservationInfo);
     }
```
34c5f255 에 반영했어용

asebn1 commented on src/main/java/roomescape/domain/Waiting.java:
```diff
@@ -0,0 +1,55 @@
+        return new Waiting(null, reservationInfo, member, rank);
+    }
+
+    public void updateRankAndReservationInfo(ReservationInfo reservationInfo) {
+        this.rank -= 1;
+        this.reservationInfo = reservationInfo;
+    }
```
항상 rank -= 1 만 하는 로직이네요
만약, 상황에따라 rank를 2 줄이는 케이스는 없거나 0으로 초기화하는 케이스는 없을까요?
외부에서 명시적으로 업데이트하는 것이 명확할 것 같아서 의견드려봅니다~

```java
public void updateRankAndReservationInfo(ReservationInfo reservationInfo, long newRank) {
    this.rank = newRank;
    this.reservationInfo = reservationInfo;
}
```

ljhee92 replied:
이 부분은 전혀 고려하지 못했던 부분이네요!
제로 덕분에 테스트하기도 편리해진 것 같습니다~!
df8673a8 에 반영했어요!

asebn1 commented on src/main/java/roomescape/application/WaitingService.java:
```diff
@@ -0,0 +1,109 @@
+        return waitingRepository.findAllByMember(member);
+    }
+
+    @Transactional
+    public void deleteWaitingById(Long id) {
+        Waiting waiting = findWaitingById(id);
+        waitingRepository.deleteById(waiting.getId());
+    }
```
본인의 대기만 삭제하도록 검증을 추가하는건 어떨까요?

ljhee92 replied:
네 좋아요!
본인의 대기만 삭제하도록 검증을 추가해주면서, 관리자는 모든 사용자의 대기를 삭제할 수 있어야 하므로 엔드포인트를 분리해주었습니다.
c3e9ab2b 에 반영했습니다!

asebn1 commented on src/test/java/roomescape/domain/WaitingTest.java:
```diff
@@ -0,0 +1,55 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+class WaitingTest {
+
+    private final LocalDateTime now = new FixedCurrentTimeService().now();
+
+    @Test
+    void 대기순위와_예약정보를_변경한다() {
+        Reservation reservation = createReservation();
+        ReservationInfo reservationInfo = ReservationInfo.create(reservation);
```
잘 작성해주셨네요 👍 
통합테스트도 추가해봅시다 :)

ljhee92 replied:
변경, 추가, 분리한 통합테스트가 많아요 제로!😭

9d771c7b, 5c69488c, d85a4db0, 1f172cfe, df1ba2ba 커밋이 통합 테스트 관련 커밋입니다!

asebn1 commented on src/main/java/roomescape/application/WaitingService.java:
```diff
@@ -0,0 +1,109 @@
+        return WaitingResponse.from(saved);
+    }
+
+    private void validateWaitingTime(ReservationInfo reservationInfo) {
+        LocalDateTime waitingDateTime = LocalDateTime.of(reservationInfo.getDate(), reservationInfo.getTime().getStartAt());
+        if (!waitingDateTime.isAfter(currentTimeService.now())) {
+            throw new IllegalArgumentException("[ERROR] 현재 시간 이후로 예약 대기할 수 있습니다.");
+        }
+    }
+
+    private void validateExistsReservationByMember(ReservationInfo reservationInfo, Member member) {
+        if (reservationRepository.existsByDateAndTimeAndThemeAndMember(reservationInfo.getDate(), reservationInfo.getTime(), reservationInfo.getTheme(), member)) {
+            throw new IllegalArgumentException("[ERROR] 이미 해당 날짜, 해당 테마, 해당 시간에 예약이 존재합니다.");
+        }
+    }
+
+    private void validateExistsWaitingByMember(Waiting waiting, Member member) {
+        if (waitingRepository.existsByReservationInfoAndMember(waiting.getReservationInfo(), member)) {
+            throw new IllegalArgumentException("[ERROR] 이미 해당 날짜, 해당 테마, 해당 시간에 예약 대기 중입니다.");
+        }
+    }
```
꼼꼼한 검증 💯

asebn1 commented on README.md:
```diff
@@ -54,3 +54,123 @@
 - [x] 화면 응답
   - [2단계 클라이언트 코드 커밋](https://github.com/woowacourse/spring-roomescape-member/commit/849391b31f2dfa359e851e95f5f2a64e21650cc8)
+
+### 3단계 - 예약 대기 기능
```
> 동시성 문제
예약을 구현할 때에도 동시성 문제는 고려하지 않았기 때문에, 이번에 대기를 구현하면서도 동시성 문제는 고려하지 않고 구현했습니다.
그래도 동시성 키워드가 크루들 사이에서 이야기가 간간히 들려와 대략적으로 어떠한 키워드(컬럼 unique 설정, 비관적 락, 낙관적 락 등)로 학습하면 좋겠다는 것만 알아두었습니다.

좋습니다 :)
말씀하신 것처럼 동시성은 실제 서비스에서 매우 중요한 문제 중 하나예요. 특히 예약이나 대기처럼 동시에 많은 사용자가 접근할 수 있는 도메인에서는 race condition 이슈가 쉽게 발생할 수 있어요.
이야기주신 키워드 위주로 학습해보면 큰 도움이 될거라 생각합니다.

asebn1 commented on src/main/java/roomescape/domain/Waiting.java:
```diff
@@ -0,0 +1,55 @@
+
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
`@NoArgsConstructor(access = AccessLevel.PROTECTED)` 설정 좋습니다.
어떤 장점이 있을까요?

ljhee92 replied:
`@NoArgsConstructor(access = AccessLevel.PROTECTED)`로 설정한다면 기본 생성자를 내부에서만 혹은 상속 관계의 클래스에서만 생성할 수 있게 제한할 수 있습니다.
기본 생성자를 외부에 public으로 열어두게 된다면 `Waiting waiting = new Waiting()`과 같이 불완전한 객체를 외부에서 생성할 수 있게 되어 객체의 일관성을 보장할 수 없게 됩니다.
private으로 설정하는 것이 외부에서 무분별하게 생성자를 생성할 수 없도록 하기 위한 가장 좋은 방법이겠지만, JPA 내부적으로 리플랙션을 사용하여 프록시 객체를 만들어야 하기 때문에 private은 사용할 수 없기 때문에 protected로 막아주었습니다.

asebn1 commented on src/main/resources/data.sql:
```diff
@@ -118,6 +118,18 @@ VALUES
+(55, CURRENT_DATE +1, 2, 9, 6, 2),
+(55, CURRENT_DATE +1, 2, 9, 7, 3),
+
+(57, CURRENT_DATE + 2, 1, 10, 3, 1),
+(57, CURRENT_DATE + 2, 1, 10, 4, 2);
```
여기도 마지막 줄 개행을 넣어볼까요?
(마지막 줄 개행은 사실상 거의 표준이라고 생각해요)

ljhee92 replied:
앗 놓친 부분이 있었네요!
934c8381 에 반영했습니다!

asebn1 reviewed (changes requested):
듀이 안녕하세요.
이번 미션 잘 개발해주셨는데요~
몇 가지 코멘트 남겼습니다. 
확인해주세요 :)

ljhee92 committed:
refactor: rank를 외부에서 명시적으로 업데이트하도록 수정

ljhee92 committed:
refactor: 예약 대기 삭제 시 본인의 예약 대기만 삭제할 수 있도록 수정

- 관리자는 모든 예약 대기를 삭제(거절)할 수 있어야 하므로 삭제 메서드를 Admin과 별도 분리

ljhee92 committed:
chore: 누락된 마지막 줄 개행 추가

ljhee92 committed:
refactor: 대기 순위 변경 시 0 이하 순위는 변경하지 못하도록 검증 추가

- 경계값 테스트 추가

ljhee92 committed:
refactor: 첫 번째 예약대기 순서를 찾을 때 존재하지 않는다면 예외 발생하도록 수정

- 사용하는 곳에서 처리해주고 있더라도 내부에서도 한 번 더 방어적으로 처리하도록 수정
- exists~ + findAll 로 처리하게 된다면 DB에 접근을 두 번 하게 되고, findAll 만으로 처리하게 된다면 예약대기가 많을 경우 효율적이지 않으므로 Optional + findTop1~ 으로 처리

ljhee92 committed:
test: 중복 테스트 코드 통합

ljhee92 committed:
test: 페이지 통합 테스트 분리

ljhee92 committed:
test: 페이지 통합 테스트 보강

- 로그인 전 혹은 권한이 없을 때 예외 케이스 추가

ljhee92 committed:
test: 데이터베이스 통합 테스트 분리

ljhee92 committed:
refactor: 예약 대기 전체 조회 관리자만 가능하도록 API 수정

ljhee92 committed:
test: 통합 테스트 보강

ljhee92 commented on src/test/java/roomescape/domain/WaitingTest.java:
```diff
@@ -0,0 +1,55 @@
+                () -> assertThat(waiting.getReservationInfo().getTime().getStartAt()).isEqualTo(now.toLocalTime().minusHours(1L)),
+                () -> assertThat(waiting.getReservationInfo().getTheme().getName()).isEqualTo("테마2")
+        );
+    }
```
c18a8910 에 경계값 테스트 추가해주었어용

ljhee92 commented:
안녕하세요 제로! 듀이입니다.
이번 리뷰도 잘 부탁드립니다😄

통합 테스트를 보강하면서, 레벨2 방탈출을 처음 시작할 때 제공되었던 `Map`을 사용하는 통합테스트 구조를 유지하려고 하다 보니 불편한 점이 많다고 느껴졌어요.
`Map`이 아닌 DTO를 사용한다면 이 불편함이 조금은 해소가 될까요..?
통합 테스트와 관련해서 어떤 쪽으로 학습을 하면 좋을지 의견이나 키워드를 던져주신다면 좋을 것 같습니다!

항상 시간 들여 리뷰해주셔서 감사합니다👍🏻

asebn1 commented on README.md:
````diff
@@ -54,3 +54,123 @@
+### 3단계 - 예약 대기 기능
+- [x] 예약 대기 요청 기능
+  - request
+    ```
````
> 통합 테스트를 보강하면서, 레벨2 방탈출을 처음 시작할 때 제공되었던 Map을 사용하는 통합테스트 구조를 유지하려고 하다 보니 불편한 점이 많다고 느껴졌어요.
Map이 아닌 DTO를 사용한다면 이 불편함이 조금은 해소가 될까요..?
통합 테스트와 관련해서 어떤 쪽으로 학습을 하면 좋을지 의견이나 키워드를 던져주신다면 좋을 것 같습니다!

Map 기반으로 통합 테스트를 유지하려다 보니 불편함이 많았다는 말에 공감합니다.
처음엔 간단하게 값을 넣고 빼기 쉬워 보여도, 시간이 지나면 어떤 값이 어떤 의미를 가지는지 명확하지 않고, 오타나 필드 누락이 있어도 컴파일 타임에 잡히지 않아서 테스트 신뢰도가 떨어질 수 있다고 생각해요.
그런 점에서 DTO를 사용하는 방식은 충분히 고려해볼 만한 대안이라고 생각해요. 구조가 명확해지고, 타입 안정성도 생기니까 유지보수나 리팩터링 할 때 훨씬 수월하죠. 현재 느끼는 불편함도 어느 정도 해소될 수 있을 거라고 봅니다.

통합 테스트 관련해서는 “어디까지를 테스트할 건지”에 대한 경계를 명확히 정하는 게 중요해요. 예를 들어 컨트롤러부터 DB까지 다 포함할 건지, 아니면 서비스까지만 테스트할 건지 등. 그 기준에 따라 테스트 설계나 도구 사용 방식도 달라지니까요.
추가로, 반복적인 테스트 데이터를 깔끔하게 관리하려면 Fixture나 TestDataBuilder 같은 패턴을 익혀두는 것도 추천드려요. 테스트 코드가 점점 길어질수록 이런 도구가 큰 도움이 된다고 생각합니다 :)

asebn1 reviewed (approved):
듀이 안녕하세요.
잘 구현해주신것 확인했습니다.
이번 미션은 이만 머지하겠습니다.
수고많으셨어요 :)
